pgxn/neon/pagestore_smgr.c

```diff
@@ -681,6 +681,33 @@ prefetch_do_request(PrefetchRequest *slot, bool *force_latest, XLogRecPtr *force
 	Assert(slot->my_ring_index == MyPState->ring_unused);
 	page_server->send((NeonRequest *) &request);
 
+	/*
+	 * Remember request LSN in the last-written LSN cache to avoid false
+	 * prefetch invalidations.
+	 *
+	 * Imagine what would happen without this, when you perform a large
+	 * sequential scan with UPDATE. The sequential scan issues a prefetch
+	 * request for each page in order, and every page is also dirtied. On
+	 * each page, the oldest page in the last-written LSN cache is evicted,
+	 * which advances the global last-written LSN. The pages being scanned are
+	 * not in the last-written cache, so each prefetch request will use the
+	 * global last-written LSN in the request and memorize that in the
+	 * slot. However, when we receive the response to the prefetch request,
+	 * the global last-written LSN has already moved forwards, and the
+	 * cross-check we make that the last-written LSN matches will fail, and we
+	 * discard the prefetched response unnecessary.
+	 *
+	 * Inserting the LSN we use in the prefetch request to the last-written LSN 
+	 * cache avoids that problem. With that, we will use the cached value in
+	 * the cross-check, instead of the more recent global last-written LSN value.
+	 */
+	SetLastWrittenLSNForBlock(
+		request.req.lsn,
+		slot->buftag.rnode,
+		slot->buftag.forkNum,
+		slot->buftag.blockNum
+	);
```

Comment:
The goal of last written LSN cache is not to return exact value when page was last updated.
We need this cache only to provide large enough gap between current WAL LSN  and request LSN (to prevent pageserver from unneeded waits). 

Yes,updating last written LSN for prefetched pages pollutes the cache. But still with current cache size (128*1024 entries)
 it will take quite long time before recently evicted from shared buffers dirty page will be also thrown away from this last written LSN cache.
 
So I believe that such simple solution is acceptable and do not cause some slowdown. 
If you do not agree, please try to create test which simulates workload where such cache pollution cause degrade of performance.

Reply:
Good catch. Agree.

But instead of moving `SetLastWrittenLSNForBlock` to `neon_prefetch` where page last written LSN is not known, I prefer just to move it inside if to the branch which handle only prefetch of subsequent pages.